pkg/agent/generate_secret_controller.go

```diff
@@ -0,0 +1,147 @@
+package agent
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	hyperv1beta1 "github.com/openshift/hypershift/api/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorapiv1 "open-cluster-management.io/api/operator/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	klusterletAnnotationFinalizer = "operator.open-cluster-management.io/hc-secret-annotation"
+	hcAnnotation                  = "create-external-hub-kubeconfig"
+)
+
+type ExternalSecretController struct {
+	hubClient   client.Client
+	spokeClient client.Client
+	log         logr.Logger
+}
+
+// var ExternalSecretPredicateFunctions = predicate.Funcs{
+// 	CreateFunc: func(e event.CreateEvent) bool {
+// 		klog.Info("KLOG TRIGERRED BY CREATE")
+// 		return true
+// 	},
+// 	UpdateFunc: func(e event.UpdateEvent) bool {
+// 		klusterletOld := e.ObjectOld.(*operatorapiv1.Klusterlet)
+// 		klusterletNew := e.ObjectNew.(*operatorapiv1.Klusterlet)
+// 		if (klusterletOld.ObjectMeta.DeletionTimestamp.IsZero() != klusterletNew.ObjectMeta.DeletionTimestamp.IsZero()) {
+// 			klog.Info("TRIGERRED BY UPDATE KLOG")
+// 			klog.Info("Old has deletion time stamp?", klusterletOld.ObjectMeta.DeletionTimestamp.IsZero())
+// 			klog.Info("New has deletion time stamp?", klusterletNew.ObjectMeta.DeletionTimestamp.IsZero())
+// 		}
+// 		return klusterletOld.ObjectMeta.DeletionTimestamp.IsZero() != klusterletNew.ObjectMeta.DeletionTimestamp.IsZero()
+// 	},
+// 	DeleteFunc: func(e event.DeleteEvent) bool {
+// 		return false
+// 	},
+// }
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *ExternalSecretController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorapiv1.Klusterlet{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
+		//WithEventFilter(ExternalSecretPredicateFunctions).
+		Complete(c)
+}
+
+// Reconcile updates the Hypershift addon status based on the Deployment status.
+func (c *ExternalSecretController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.log.Info(fmt.Sprintf("reconciling annotation for %s", req.Name))
+	defer c.log.Info(fmt.Sprintf("done reconciling annotation for %s", req.Name))
+
+	if !strings.Contains(req.Name, "klusterlet-") {
+		c.log.Info("klusterlet not from a managed cluster")
+		return ctrl.Result{}, nil //No need to error
+	}
+
+	//Get klusterlet
+	klusterletNamespaceNsn := types.NamespacedName{
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}
+	klusterlet := &operatorapiv1.Klusterlet{}
+	if err := c.spokeClient.Get(ctx, klusterletNamespaceNsn, klusterlet); err != nil {
+		c.log.Error(err, "Unable to get klusterlet")
```

Comment:
Lowercase "u" for unable. Do we want to include the klusterlet ns and name in the error msg?